typing/typecore.ml

```diff
@@ -990,13 +968,133 @@ let rec has_literal_pattern p = match p.ppat_desc with
+   non-termination (this is a generic proof-search problem) or
+   compilation-time slowdowns.
+
+   [explosion_fuel] is this depth limit on wildcard explosion. *)
+and explosion_fuel = int
+
+
+(** A counter-example pattern may not be well-typed because the
+   generator does not know about GADTs. In particular, it may generate
+   (even in depth) disjunctions of the form
+
+         (p1 | p2 | ... | pn),
+
+   where some of the pi may in fact be ill-typed. If at least one of
+   them is well-typed, we do have a valid counter-example. So we want
+   the type-checker to succeed if at least one of the alternatives is
+   well-typed.
+
+   Note that the choice of value in such an or-pattern can influence
+   the validity of other parts of the counter-example: if the
+   alternatives are GADT constructors, each one may add equations that
+   restrict other part of the counter-example in its own way. As
+   a consequence, evaluating each choice may require arbitrary
+   search/filtering in other parts of the counter-example (in
+   implementation terms, the continuation [k] is called on each
+   alternative).
+
+   Searching after each alternative in this way may be very costly. It
+   is only useful when GADTs are involved inside the or-pattern; if no
+   GADT constructors occur (possibly deep) inside an alternative, it
+   does no restrict the rest of the search space, so we can go with
+   this alternative and never try the others. But we don't know
+   whether a GADT constructor will occur (we haven't checked the
+   pattern yet!).
+
+   A possible implementation is to optimistically assume that
+   splitting will not be necessary, and split only if we hit a GADT
+   under an or-pattern. This strategy is faster when GADTs are rare.
+
+   The [splitting_mode] setting determines the splitting policy.  *)
+and splitting_mode =
+  | Pessimistic
+  (** Always split or-patterns
+
+     We use this mode when generating exhaustivity counter-examples
+     to check a refutation clause (p -> .), which typically involves
+     GADTs. *)
+  | Optimistic of { inside_nonsplit_or: bool; }
+  (** Split or-patterns on-demand.
+
+     In this mode we optimistically assume that splitting will not be
+     necessary; we first type-check alternatives under an or-pattern
+     without splitting. If we then hit a GADT constructor, we jump back
+     to the or-pattern and fallback to the splitting implementation.
+
+     We use this mode when checking exhaustivity of pattern matching;
+     typical OCaml code uses few GADT constructors on average. *)
+
+
+(** The exhaustivity checker produces counter-examples that contain
+   synthetic/fake variant constructors and record labels, as a hackish
+   way to encode a Typedtree.pattern back into a Parsetree.pattern
+   without losing information. The tables below are used to read back
+   the proper typed constructor/label from the synthetic parsetree-label. *)
+and counter_example_synthetic_names = {
```

Comment:
Why not inline this one in its parent?

Reply:
I guess Gabriel wanted to be regular about where the documentation is placed.

I'm fine with this particular type, but have no objection to it being inlined in its parents.